### ISSUE TITLE
Type narrowing in nested loop evaluates incorrect type

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#3987]
A recent change for [this bug](https://github.com/microsoft/pyright/issues/3985) uncovered a previously-masked type narrowing issue that is seen in the example below.

```python
from concurrent import futures
from concurrent.futures import Future
from typing import Any, Dict, Optional

class A:
    def __init__(self):
        self.pending: Optional[Dict[Future[Any], int]]
        self.foo: bool

    def poll(self):
        assert self.pending is not None
        while True:
            if self.pending:
                pass

            ready, _ = futures.wait(self.pending)

            for future_id in ready:
                self.pending.pop(future_id) # False negative error on this line

                future_id.result()
                if self.foo:
                    pass
```

**Comment by erictraut** [comment]
This issue is addressed in pyright 1.1.273, which I just published. It will also be included in a future release of pylance.
